webhooks:  Fix reentrant protection

Arksine opened the pull request:
This fixes a bug where pending requests would be dropped when the ServerConnection is "busy".  The "emergency_stop" request is now checked out of order and will be processed if it is properly decoded and validated.

Signed-off-by:  Eric Callahan <arksine.code@gmail.com>

KevinOConnor commented:
Thanks.  If possible, I'd prefer not to introduce the out-of-order command check - that's a really ugly g-code hack.  For api access, I'd guess that we would want to run each command and not wait for it to complete.  That is, if a particular command calls `reactor.pause()` (or calls something that in turn calls `reactor.pause()`) then we'd want to suspend that command but go on to run the next command.  If this was done, I don't think we'd need special logic to handle shutdown requests.  What do you think?

-Kevin

Arksine commented:
That sounds good.  It seems to me that the easiest way to approach this is to simply execute the web request handlers in a reactor callback?  I reverted the previous commit and pushed a patch that does something like this.  This should guarantee that _process_received() does not block so I also removed the logic that checked for reentrancy.

Arksine commented:
I went ahead and added another commit to this that switches the check that disables the server from debugoutput to debuginput.  I'm currently writing regression tests for Moonraker and it occurred to me that it would be extremely useful to be able connect and test against Klippy when "debugoutput" is enabled.

KevinOConnor commented:
Thanks.  Looks good to me.  Did you want me to squash these into one commit?  Otherwise, it's probably best to rebase so we're not committing and then reverting the old approach.

-Kevin

Arksine commented:
Great, I went ahead and rebased for you.  Thanks!

KevinOConnor commented:
Thanks.

-Kevin
